### PR TITLE
Provide error message on return value is not a httpx response instance

### PR DESCRIPTION
**Pull request by tzing**
We already have a check for response type via router.py L257

```py
    # Mocked response
    assert isinstance(mock, httpx.Response)
    response = mock
```

But this `AssertionError` would be later catched by `Mocker.handler` in

```py
    def handler(cls, httpx_request):
        httpx_response = None
        error = None
        for router in cls.routers:
            try:
                httpx_response = router.handler(httpx_request)
            except AssertionError as e:
                error = e.args[0]  # <-- HERE
                continue
            else:
                break
        else:
            assert httpx_response, error
        return httpx_response
```

However, we do not provide a error message and thus it could not get `e.args[0]` and would raise `IndexError`

```py
    @classmethod
    def handler(cls, httpx_request):
        httpx_response = None
        error = None
        for router in cls.routers:
            try:
                httpx_response = router.handler(httpx_request)
            except AssertionError as e:
>               error = e.args[0]
E               IndexError: tuple index out of range

.venv/lib/python3.9/site-packages/respx/mocks.py:88: IndexError
```

This PR add message to it so it will show corresponding error for user.

### CODE REVIEW

**Comment by lundberg** [comment]
Good point @tzing and thanks for contributing! ....but 😉 

The only reason why this `assert isinstance(mock, httpx.Response)` is there is unfortunately due to "help" `mypy`.

If you look at the `Router.match` method, the return value is typed as...

`-> Tuple[Optional[Route], Optional[Union[httpx.Request, httpx.Response]]]`

...resulting in `mock` var in the `resolve` method is ensured to be a request, response or None. 

The [if-elif-else](https://github.com/lundberg/respx/blob/86d1c7f73bbeaa50cc025ad395d7797e52017066/respx/router.py#L245-L259)  block was hard to write in an other way with proper typing making mypy happy.

We could change the `else:` to an `elif isinstance(mock, httpx.Response):` and remove the `assert`, but `mypy` will then complain about missing `else:` fallback, so we need to add that, but what to do there, raise some error that never could happen, i.e. can't be tested and get coverage on 😉  

I'm OK with adding this assertion error message, but I'd rather see us get rid of this assert completely if possible.

**Comment by tzing** [comment]
Sorry i didn't get it

On assertion failed, it will raised `AssertionError` and breaks current stack.
Why does it have impact on return type?

Yet another question to this, I've just trying `mypy respx` on this branch and it does not gives any error or warning.
Do I miss anything? I've never use mypy.

**Comment by lundberg** [comment]
> On assertion failed, it will raised `AssertionError` and breaks current stack.
> Why does it have impact on return type?

The row `assert isinstance(mock, httpx.Response)` is just a typing "hack" to please `mypy`.

The `assert` could be changed to an `elif`, but that would "upset" mypy. Reason why I mentioned the return type of `.match()` was to show that `mock` can only be request, response or None.

> Yet another question to this, I've just trying `mypy respx` on this branch and it does not gives any error or warning.
> Do I miss anything? I've never use mypy.

Nope, the error message is OK and doesn't affect mypy.

So my reasoning is more about that we should/could get rid of the assert instead of adding the error message. To try this, you can remove the assert row and run mypy and you'll see. Also you can try the elif change and run mypy.

**Comment by lundberg** [comment]
Gave this another shot locally and this would both satisfy mypy, without any change of logic, and get rid of the assert...

```py
if route is None:
    # Assert we always get a route match, if check is enabled
    assert not self._assert_all_mocked, f"RESPX: {request!r} not mocked!"

    # Auto mock a successful empty response
    response = httpx.Response(200)

elif isinstance(mock, httpx.Response):
    # Mocked response
    response = mock
```

The removed `elif mock == request:` wasn't needed since the response is set to `None` by default earlier.

File: respx/router.py [hunk 1]
```diff
@@ -253,11 +253,12 @@ def resolve(self, request: httpx.Request) -> Optional[httpx.Response]:
...
                 response = None
 
-            else:
-                # Mocked response
-                assert isinstance(mock, httpx.Response)
+            elif isinstance(mock, httpx.Response):
                 response = mock
 
+            else:
+                raise TypeError(f"RESPX: {mock!r} is not a instance of httpx.Response")
```
Comment (lundberg) [inline respx/router.py:260]: 😉 
```suggestion
                raise TypeError(f"RESPX: {mock!r} is not an instance of httpx.Response")
```

File: tests/test_router.py [hunk 2]
```diff
@@ -215,6 +216,10 @@ def test_side_effect_exception():
     with pytest.raises(httpcore.NetworkError):
         router.resolve(request)
 
+    request = httpx.Request("GET", "https://hot.dog")
+    with pytest.raises(TypeError):
+        router.resolve(request)
```
Comment (lundberg) [inline tests/test_router.py:221]: Please move this from the the exception side effect test to a separate test case, e.g. something like `test_resolve_invalid_route_return_value`.

**Review by lundberg — COMMENTED** [review]
Looks better 👍 

Although, it got me thinking that we should probably guard for setting the wrong `return_value` in the first place, e.g. the `TypeError` should also (only?) be checked and raised in the `Route.return_value` setter.

Same goes for the `ValueError` in `Route._call_side_effect`, which also should be a `TypeError`.

File: respx/router.py (hunk 1 above)
Comment (lundberg) [reply respx/router.py:260]: btw, this line will never be reached if we type check in the `Route.return_value` setter as I mention 😉

**Comment by tzing** [comment]
😉 Well I thought it's more intuitively to check type in `Route.return_value` setter.
Shall we keep going on this PR? I'd prefer to create a new one for setter.

**Comment by lundberg** [comment]
> Shall we keep going on this PR? I'd prefer to create a new one for setter.

We should merge your #133 first, then once this PR is rebased we can discuss how to proceed with this one, since the new test will probably fail / not reach.

**Comment by lundberg** [comment]
Closing this for now, please re-open if needed.

**Pull request closed without merging**